src/cryptography/hazmat/backends/openssl/decode_asn1.py

```diff
@@ -24,9 +24,23 @@
 def _obj2txt(backend, obj):
     # Set to 80 on the recommendation of
     # https://www.openssl.org/docs/crypto/OBJ_nid2ln.html#return_values
+    #
+    # But OIDs longer than this occur in real life (e.g. Active
+    # Directory makes some very long OIDs).  So we need to detect
+    # and properly handle the case where the default buffer is not
+    # big enough.
+    #
     buf_len = 80
     buf = backend._ffi.new("char[]", buf_len)
+
+    # 'res' is the number of bytes that *would* be written if the
+    # buffer is large enough.  If 'res' > buf_len - 1, we need to
+    # alloc a big-enough buffer and go again.
     res = backend._lib.OBJ_obj2txt(buf, buf_len, obj, 1)
+    if res > buf_len - 1:  # account for terminating null byte
+        buf_len = res + 1
```

Comment:
The fix is not sufficient. ``OBJ_obj2txt(buf, buf_len, obj, 1)`` does **not** return the actual required buffer size. You have to call the function with NULL buffer to make it calculate the correct size: ``OBJ_obj2txt(NULL, 0, obj, 1)``.